### PR TITLE
[Metrics SDK] Provide builders to avoid exposing SDK internals

### LINKED ISSUE #1485: [Metrics SDK] Provide builders to avoid exposing SDK internals
Same as issue #1393, but for Metrics.

Please assign to me.

### PR DESCRIPTION
**Pull request by marcalff**
Fixes #1485 

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by marcalff** [comment]
Now ready for review.

File: sdk/include/opentelemetry/sdk/metrics/meter_provider_factory.h [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  to make additional calls to the SDK class, until such a time when
+  the builders can take all the necessary input at once,
+  for example using a std::vector<MetricReader> to add all readers.
+
+  Implementing (2) is forcing technical debt, and prevents the
+  calling application configuring the SDK to be decoupled from it,
+  making deployment of shared libraries much more difficult.
+
+  The design choice here is to return (1) an API MeterProvider,
+  even if this forces, temporarily, existing applications to use a downcast.
```
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/metrics/meter_provider_factory.h:56]: Yes, option 1 would be the right choice. We can have a separate PR for overloaded `sdk::MeterProvider` constructor taking both `Views` and `MetricReader` as an argument. And modify existing examples accordingly.

File: examples/metrics_simple/metrics_ostream.cc [hunk 2]
```diff
@@ -27,59 +33,75 @@ namespace
...
+  std::shared_ptr<metrics_sdk::AggregationConfig> aggregation_config(
+      std::move(histogram_aggregation_config));
+
+  auto histogram_view = metrics_sdk::ViewFactory::Create(
+      name, "description", metrics_sdk::AggregationType::kHistogram, aggregation_config);
+
   p->AddView(std::move(histogram_instrument_selector), std::move(histogram_meter_selector),
              std::move(histogram_view));
+
+  std::shared_ptr<opentelemetry::metrics::MeterProvider> provider(std::move(u_provider));
```
Comment (lalitb) [inline examples/metrics_simple/metrics_ostream.cc:104]: These multiple casts from `api:::provider` -> `sdk::provider` -> `api::provider` are bit confusing, but seems we don't have better choice for now.

File: examples/otlp/BUILD [hunk 3]
```diff
@@ -93,3 +93,22 @@ cc_binary(
...
+        "otlp",
+    ],
+    deps = [
+        "//api",
+        "//examples/common/metrics_foo_library:common_metrics_foo_library",
+        "//exporters/otlp:otlp_http_exporter",
+        "//exporters/otlp:otlp_http_metric_exporter",
+        "//sdk/src/metrics",
+    ],
+)
```
Comment (lalitb) [inline examples/otlp/BUILD:114]: Thanks for adding OTLP HTTP metric exporter example :)

File: examples/otlp/grpc_metric_main.cc [hunk 4]
```diff
@@ -25,24 +28,31 @@ namespace otlp_exporter = opentelemetry::exporter::otlp;
...
+  reader_options.export_interval_millis = std::chrono::milliseconds(1000);
+  reader_options.export_timeout_millis  = std::chrono::milliseconds(500);
+
+  auto reader =
+      metric_sdk::PeriodicExportingMetricReaderFactory::Create(std::move(exporter), reader_options);
+
+  auto context = metric_sdk::MeterContextFactory::Create();
+  context->AddMetricReader(std::move(reader));
+
+  // FIXME: MeterProvider / MeterProviderFactory should take a unique_ptr
```
Comment (lalitb) [inline examples/otlp/grpc_metric_main.cc:51]: What is this FIXME about as the implementation is similar to TracerProvider. MeterContext is used across each of the Meters, so passed as shared_ptr.

**Review by lalitb — APPROVED** [review]
Nicely done, Thanks for the PR.

File: examples/otlp/grpc_metric_main.cc (hunk 4 above)
Comment (marcalff) [reply examples/otlp/grpc_metric_main.cc:51]: I filed issues #2203 and #2204 for this, the point is that due to spec clarifications, the context should precisely not be shared between providers, for both traces and metrics.

Comment (marcalff) [reply examples/otlp/grpc_metric_main.cc]: Cleaned up the FIXME comments, will be addressed separately by the new issues filed.

**Pull request merged**